tests/__init__.py

```diff
@@ -0,0 +1,16 @@
+from functools import lru_cache
+
+from dwave.cloud import Client
+from dwave.cloud.exceptions import ConfigFileError
+
+@lru_cache(maxsize=None)
+def qpu_available():
+    """Check whether QPU solver is available"""
+    try:
+        with Client.from_config() as client:
```

Comment:
This is highly dependent on user's environment (but it works).

Not sure what's the intended use case for this example, but we could also for example:

1) use a local config file (what we currently do in the cloud-client):
```
Client.from_config(config_file='/path/to/local/test.conf')
```

2) ignore the config file(s) completely:
```
Client.from_config(config_file=False)
```
effectively requiring the token to be provided via env (`DWAVE_API_TOKEN` env var).

I prefer option 2, with test running instructions added to readme:
```
DWAVE_API_TOKEN='my-token' python -m unittest ....
```

Reply:
Than again, we use `from_config` in the demo anyway, and your implementation is closest to that, so it makes most sense in the test.

Reply:
To my understanding, the issue that we're trying to address is that the Circle CI test runners do not have access to an API token, so we want to be able to gracefully skip tests that require the QPU when the token is not available.  I don't see how approach 1 would address this, and I don't understand the advantage in approach 2 of specifying `config_file=False`.  Doesn't that just add an extra step when running the tests with an API token?

Ultimately, the code that creates the sampler instance is going to to eventually call into `Client.from_config` anyway, right?  My thinking was to just try and replicate that similar logic within the `qpu_available` function so that it can be checked prior and used to skip tests if necessary.

https://github.com/dwave-examples/structural-imbalance/blob/a51dad8b170cb39494884ed4b93d31c87be5d92a/dwave_structural_imbalance_demo/interfaces.py#L89

https://github.com/dwavesystems/dwave-system/blob/32e9065cddeb123106b43d947877043c1a2ccc78/dwave/system/samplers/dwave_sampler.py#L163-L164

Reply:
Yes, that's exactly why I posted my second comment above. Initially I was approaching it like this was a library, not demo code.

FWIW, both options would work equally well on the CI, and both would require the token in env (what we do anyway). But they would also require this additional step for the user, which is not ideal.

What's good about ignoring user's config during tests, though, is we have better control over test runs (i.e. better reproducibility). Again, not a big deal for a demo.